{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "09f33caa-3ddc-4c8b-ad09-fb0e3222edea",
   "metadata": {},
   "source": [
    "### Load needed libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "37dce2f1-2427-4808-938f-e51cd8cc5d4f",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "import os\n",
    "import numpy as np\n",
    "import anndata\n",
    "import re\n",
    "import pandas as pd\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.colors as colors\n",
    "import json\n",
    "from joblib import Parallel, delayed\n",
    "from joblib import parallel_backend\n",
    "import glob\n",
    "%matplotlib inline\n",
    "import csv\n",
    "import scanpy as sc\n",
    "import copy\n",
    "import random\n",
    "import scvi\n",
    "import scipy as sp\n",
    "import warnings\n",
    "from datetime import datetime\n",
    "from igraph import *\n",
    "from pylab import *\n",
    "\n",
    "sc.settings.n_jobs = 32\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "plt.ioff()\n",
    "\n",
    "pwd = os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f191cc77-08bc-472f-85e5-ddbdd62cfc03",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prepare_anndata(adata, region, split_key, split_value, date_tag, metadata_to_color):\n",
    "    if os.path.exists(os.path.join(pwd, \"output\", region, split_value.replace(\"/\", \" \") + \"_CZI_display.\" + date_tag + \".h5ad\")):\n",
    "        print(str(datetime.datetime.now()) + \" -- Skipping \" + split_value + \" because an output AnnData exists\")\n",
    "        return\n",
    "        \n",
    "    print(str(datetime.datetime.now()) + \" -- Starting export of \" + split_value)\n",
    "    primary_data = True\n",
    "    if split_value != \"Whole Taxonomy\":\n",
    "        markers = pd.read_csv(os.path.join(pwd, \"input\", region, \"models\", split_value.replace(\"/\", \" \"), \"scVI_model\", \"var_names.csv\"), header=None)\n",
    "        markers = markers[0].to_list()\n",
    "        model = scvi.model.SCVI.load(os.path.join(pwd, \"input\", region, \"models\", split_value.replace(\"/\", \" \"), \"scVI_model\"), adata[:, markers])\n",
    "        with parallel_backend('threading', n_jobs=32):\n",
    "            adata.obsm[\"X_scVI\"] = model.get_latent_representation()\n",
    "            sc.pp.neighbors(adata, use_rep=\"X_scVI\")\n",
    "            sc.tl.umap(adata)\n",
    "        primary_data = False\n",
    "\n",
    "    # Add uns elements required by the schema\n",
    "    adata.uns[\"schema_version\"] = \"3.0.0\"\n",
    "    adata.uns[\"title\"] = split_value + \" - \" + region + \": Seattle Alzheimer's Disease Atlas (SEA-AD)\"\n",
    "    #adata.uns[\"X_normalization\"] = \"ln(UP10K+1)\"\n",
    "    adata.uns[\"batch_condition\"] = [\"Specimen ID\"]\n",
    "    adata.uns[\"default_embedding\"] = \"X_umap\"\n",
    "    del adata.uns[\"_scvi\"]\n",
    "    del adata.uns[\"subclass_scANVI_colors\"]\n",
    "    del adata.uns[\"supertype_scANVI_leiden_colors\"]\n",
    "    \n",
    "    # Remove unwanted obsm values\n",
    "    del adata.obsm[\"_scvi_extra_categoricals\"]\n",
    "    del adata.obsm[\"_scvi_extra_continuous\"]\n",
    "    \n",
    "    # Convert var_names to ensembl IDs\n",
    "    ensembl_ids = pd.read_csv(os.path.join(pwd, \"input\", \"genes.gtf\"), sep=\"\\t\", header=None, skiprows=5)\n",
    "    re_name = re.compile('gene_name \"([^\"]+)')\n",
    "    re_id = re.compile('gene_id \"([^\"]+)')\n",
    "    gene_names = [re_name.search(ensembl_ids.iloc[i, 8])[1] for i in range(ensembl_ids.shape[0])]\n",
    "    gene_ids = [re_id.search(ensembl_ids.iloc[i, 8])[1] for i in range(ensembl_ids.shape[0])]\n",
    "    names_to_ids = pd.DataFrame(\n",
    "        {\n",
    "            \"gene_ids\": gene_ids,\n",
    "            \"gene_names\": gene_names\n",
    "        }\n",
    "    )\n",
    "    names_to_ids = names_to_ids.drop_duplicates()\n",
    "    names_to_ids.index = names_to_ids[\"gene_ids\"]\n",
    "    \n",
    "    # Compare ensemblIDs from 10x (Gencode v32) to CZI (Gencode v38) and recover 15 IDs with the same name\n",
    "    gencode_ids = pd.read_csv(os.path.join(pwd, \"input\", \"gencode.v38.primary_assembly.annotation.gtf\"), sep=\"\\t\", header=None, skiprows=5)\n",
    "    re_name = re.compile('gene_name \"([^\"]+)')\n",
    "    re_id = re.compile('gene_id \"([^\"]+)')\n",
    "    gencode_names = [re_name.search(gencode_ids.iloc[i, 8])[1] for i in range(gencode_ids.shape[0])]\n",
    "    gencode_ids = [re_id.search(gencode_ids.iloc[i, 8])[1] for i in range(gencode_ids.shape[0])]\n",
    "    gencode_names_to_ids = pd.DataFrame(\n",
    "        {\n",
    "            \"gene_ids\": gencode_ids,\n",
    "            \"gene_names\": gencode_names\n",
    "        }\n",
    "    )\n",
    "    gencode_names_to_ids = gencode_names_to_ids.drop_duplicates()\n",
    "    gencode_names_to_ids[\"gene_ids\"] = [i.split(\".\")[0] for i in gencode_names_to_ids[\"gene_ids\"]]\n",
    "    gencode_names_to_ids.index = gencode_names_to_ids[\"gene_names\"]\n",
    "    tossed_genes = names_to_ids.loc[np.setdiff1d(names_to_ids[\"gene_ids\"], gencode_names_to_ids[\"gene_ids\"]), \"gene_names\"]\n",
    "    names_to_ids.index = names_to_ids[\"gene_names\"].copy()\n",
    "    names_to_ids.loc[np.intersect1d(tossed_genes, gencode_names_to_ids.index), :] = gencode_names_to_ids.loc[np.intersect1d(tossed_genes, gencode_names_to_ids.index), :]\n",
    "    \n",
    "    # Recover IDs where gene to ID mapping was not 1-to-1\n",
    "    names_to_ids = names_to_ids.loc[~(names_to_ids[\"gene_names\"].duplicated(keep=\"first\")), :]\n",
    "    no_match_gene = np.setdiff1d(adata.var_names, names_to_ids.gene_names)\n",
    "    tmp = pd.DataFrame(\n",
    "        {\n",
    "            \"gene_ids\": [i.split(\" \")[1] for i in no_match_gene],\n",
    "            \"gene_names\": no_match_gene\n",
    "        }\n",
    "    )\n",
    "    names_to_ids = pd.concat([names_to_ids, tmp], axis=0)\n",
    "    names_to_ids.index = names_to_ids[\"gene_names\"].copy()\n",
    "    adata.var_names = pd.Index(names_to_ids.loc[adata.var_names, \"gene_ids\"])\n",
    "    \n",
    "    # Add var elements required by the schema\n",
    "    #adata.var[\"feature_biotype\"] = \"gene\"\n",
    "    adata.var.drop([\"gene_ids\"], axis=1, inplace=True)\n",
    "    \n",
    "    # add obs elements required by the schema\n",
    "    adata.obs_names = adata.obs[\"exp_component_name\"].copy()\n",
    "    \n",
    "    adata.obs[\"donor_id\"] = adata.obs[\"donor_name\"].copy()\n",
    "    \n",
    "    adata.obs[\"assay_ontology_term_id\"] = adata.obs[\"method\"].copy()\n",
    "    metadata_to_ids = {\n",
    "        \"10Xv3\": \"EFO:0009922\",\n",
    "        \"10Xv3.1\": \"EFO:0009922\",\n",
    "        \"10xMulti\": \"EFO:0030059\"   # Old: EFO:0008995\n",
    "    }\n",
    "    adata.obs[\"assay_ontology_term_id\"] = adata.obs[\"assay_ontology_term_id\"].map(metadata_to_ids).astype('category')\n",
    "    \n",
    "    adata.obs[\"suspension_type\"] = \"nucleus\"\n",
    "    \n",
    "    adata.obs[\"cell_type_ontology_term_id\"] = adata.obs[\"subclass_scANVI\"].copy()\n",
    "    metadata_to_ids = {\n",
    "        \"L2/3 IT\": \"CL:4023040\",\n",
    "        \"L4 IT\": \"CL:4023040\",\n",
    "        \"L5 IT\": \"CL:4023040\",\n",
    "        \"L6 IT\": \"CL:4023040\",\n",
    "        \"L6 IT Car3\": \"CL:4023040\",\n",
    "        \"L6b\": \"CL:4023038\",\n",
    "        \"L5 ET\": \"CL:4023041\",\n",
    "        \"L5/6 NP\": \"CL:4023012\",\n",
    "        \"L6 CT\": \"CL:4023013\",\n",
    "        \"Lamp5\": \"CL:4023011\",\n",
    "        \"Lamp5 Lhx6\": \"CL:4023011\",# **\n",
    "        \"Sncg\": \"CL:4023015\",\n",
    "        \"Vip\": \"CL:4023016\",\n",
    "        \"Pax6\": \"CL:4023070\",\n",
    "        \"Sst\": \"CL:4023017\",\n",
    "        \"Sst Chodl\": \"CL:4023017\",\n",
    "        \"Pvalb\": \"CL:4023018\",\n",
    "        \"Chandelier\": \"CL:4023036\",\n",
    "        \"Astrocyte\": \"CL:0002605\",# **\n",
    "        \"Endothelial\": \"CL:1001602\", # **\n",
    "        \"Microglia-PVM\": \"CL:0000129\", # **\n",
    "        \"OPC\": \"CL:0002453\",\n",
    "        \"Oligodendrocyte\": \"CL:0000128\", # **\n",
    "        \"VLMC\": \"CL:4023051\"\n",
    "    }\n",
    "    adata.obs[\"cell_type_ontology_term_id\"] = adata.obs[\"cell_type_ontology_term_id\"].map(metadata_to_ids).astype('category')\n",
    "    \n",
    "    adata.obs[\"development_stage_ontology_term_id\"] = [str(i) for i in adata.obs[\"age_at_death\"]]\n",
    "    metadata_to_ids = {\n",
    "        \"29.0\": \"HsapDv:0000123\",\n",
    "        \"42.0\": \"HsapDv:0000136\",\n",
    "        \"50.0\": \"HsapDv:0000144\",\n",
    "        \"68.0\": \"HsapDv:0000162\",\n",
    "        \"69.0\": \"HsapDv:0000163\",\n",
    "        \"70.0\": \"HsapDv:0000164\",\n",
    "        \"72.0\": \"HsapDv:0000166\",\n",
    "        \"75.0\": \"HsapDv:0000169\",\n",
    "        \"77.0\": \"HsapDv:0000171\",\n",
    "        \"78.0\": \"HsapDv:0000172\",\n",
    "        \"80.0\": \"HsapDv:0000206\",\n",
    "        \"81.0\": \"HsapDv:0000207\",\n",
    "        \"82.0\": \"HsapDv:0000208\",\n",
    "        \"83.0\": \"HsapDv:0000209\",\n",
    "        \"84.0\": \"HsapDv:0000210\",\n",
    "        \"85.0\": \"HsapDv:0000211\",\n",
    "        \"86.0\": \"HsapDv:0000212\",\n",
    "        \"87.0\": \"HsapDv:0000213\",\n",
    "        \"88.0\": \"HsapDv:0000214\",\n",
    "        \"89.0\": \"HsapDv:0000215\",\n",
    "        \"90.0\": \"HsapDv:0000095\",\n",
    "        \"91.0\": \"HsapDv:0000095\",\n",
    "        \"92.0\": \"HsapDv:0000095\",\n",
    "        \"93.0\": \"HsapDv:0000095\",\n",
    "        \"94.0\": \"HsapDv:0000095\",\n",
    "        \"95.0\": \"HsapDv:0000095\",\n",
    "        \"96.0\": \"HsapDv:0000095\",\n",
    "        \"97.0\": \"HsapDv:0000095\",\n",
    "        \"98.0\": \"HsapDv:0000095\",\n",
    "        \"99.0\": \"HsapDv:0000095\",\n",
    "        \"100.0\": \"HsapDv:0000095\",\n",
    "        \"102.0\": \"HsapDv:0000095\"\n",
    "    }\n",
    "    adata.obs[\"development_stage_ontology_term_id\"] = adata.obs[\"development_stage_ontology_term_id\"].map(metadata_to_ids).astype('category')\n",
    "    \n",
    "    adata.obs[\"disease_ontology_term_id\"] = \"PATO:0000461\"\n",
    "    adata.obs.loc[adata.obs[\"ch_cognitivestatus_binary\"] == \"Dementia\", \"disease_ontology_term_id\"] = \"MONDO:0001627\"\n",
    "    adata.obs[\"disease_ontology_term_id\"] = adata.obs[\"disease_ontology_term_id\"].astype(\"category\")\n",
    "    \n",
    "    adata.obs[\"self_reported_ethnicity_ontology_term_id\"] = \"unknown\"\n",
    "    adata.obs.loc[(adata.obs[\"ch_race___1\"] == \"Checked\")\n",
    "              & (adata.obs[\"self_reported_ethnicity_ontology_term_id\"] == \"unknown\"), \"self_reported_ethnicity_ontology_term_id\"] = \"HANCESTRO:0005\"\n",
    "    adata.obs.loc[(adata.obs[\"ch_race___2\"] == \"Checked\")\n",
    "              & (adata.obs[\"self_reported_ethnicity_ontology_term_id\"] == \"unknown\"), \"self_reported_ethnicity_ontology_term_id\"] = \"HANCESTRO:0016\"\n",
    "    adata.obs.loc[(adata.obs[\"ch_race___3\"] == \"Checked\")\n",
    "              & (adata.obs[\"self_reported_ethnicity_ontology_term_id\"] == \"unknown\"), \"self_reported_ethnicity_ontology_term_id\"] = \"HANCESTRO:0008\"\n",
    "    \n",
    "    adata.obs.loc[(adata.obs[\"donor_name\"] == \"H200.1023\"), \"self_reported_ethnicity_ontology_term_id\"] = \"HANCESTRO:0015\"\n",
    "    adata.obs.loc[(adata.obs[\"ch_raceother\"] == \"Mixed\"), \"self_reported_ethnicity_ontology_term_id\"] = \"HANCESTRO:0306\"\n",
    "    \n",
    "    # All other remaining reference donors were Caucasian\n",
    "    adata.obs.loc[(adata.obs[\"self_reported_ethnicity_ontology_term_id\"] == \"unknown\"), \"self_reported_ethnicity_ontology_term_id\"] = \"HANCESTRO:0005\"\n",
    "    \n",
    "    adata.obs[\"self_reported_ethnicity_ontology_term_id\"] = adata.obs[\"self_reported_ethnicity_ontology_term_id\"].astype(\"category\")\n",
    "\n",
    "    adata.obs[\"organism_ontology_term_id\"] = \"NCBITaxon:9606\"\n",
    "    adata.obs[\"organism_ontology_term_id\"] = adata.obs[\"organism_ontology_term_id\"].astype(\"category\")\n",
    "    \n",
    "    adata.obs[\"sex_ontology_term_id\"] = adata.obs[\"sex\"].copy()\n",
    "    adata.obs[\"sex_ontology_term_id\"].cat.rename_categories(\n",
    "        {\n",
    "            \"Female\": \"PATO:0000383\",\n",
    "            \"Male\": \"PATO:0000384\"\n",
    "        },\n",
    "        inplace=True\n",
    "    )\n",
    "    adata.obs.drop(\"sex\", axis=1, inplace=True)\n",
    "    \n",
    "    adata.obs[\"tissue_ontology_term_id\"] = \"UBERON:0002771\"\n",
    "    adata.obs[\"tissue_ontology_term_id\"] = adata.obs[\"tissue_ontology_term_id\"].astype(\"category\")\n",
    "    \n",
    "    adata.obs[\"is_primary_data\"] = primary_data\n",
    "    \n",
    "    # Move UMIs to adata.raw and freeze it\n",
    "    adata = adata[:, np.intersect1d(adata.var_names, gencode_names_to_ids[\"gene_ids\"])].copy()\n",
    "    tmp = adata.X.copy()\n",
    "    adata.X = adata.layers[\"UMIs\"]\n",
    "    del adata.layers[\"UMIs\"]\n",
    "    adata.raw = adata\n",
    "    adata.X = tmp\n",
    "    del tmp\n",
    "    adata.var[\"feature_is_filtered\"] = False\n",
    "\n",
    "    \n",
    "    # Metadata tweaks for public display\n",
    "    adata.obs.loc[adata.obs[\"uwa\"] == \"UWA 7186\", \"sex_ontology_term_id\"] = \"PATO:0000384\"\n",
    "    \n",
    "    adata.obs[\"age_at_death_binned\"] = \"Less than 65 years old\"\n",
    "    adata.obs.loc[(adata.obs[\"age_at_death\"] >= 65) & (adata.obs[\"age_at_death\"] < 78), \"age_at_death_binned\"] = \"65 to 77 years old\"\n",
    "    adata.obs.loc[(adata.obs[\"age_at_death\"] >= 78) & (adata.obs[\"age_at_death\"] < 90), \"age_at_death_binned\"] = \"78 to 89 years old\"\n",
    "    adata.obs.loc[(adata.obs[\"age_at_death\"] >= 90), \"age_at_death_binned\"] = \"90+ years old\"\n",
    "\n",
    "    adata.obs[\"age_at_death_binned\"] = adata.obs[\"age_at_death_binned\"].astype(\"category\")\n",
    "    \n",
    "    adata.obs[\"ch_cognitivestatus_binary\"].cat.add_categories(\"Reference\", inplace=True)\n",
    "    adata.obs[\"ch_cognitivestatus_binary\"].fillna(\"Reference\", inplace=True)\n",
    "\n",
    "    adata.obs[\"adneurochange\"].cat.add_categories(\"Reference\", inplace=True)\n",
    "    adata.obs[\"adneurochange\"].fillna(\"Reference\", inplace=True)\n",
    "\n",
    "    adata.obs[\"braak\"].cat.add_categories(\"Reference\", inplace=True)\n",
    "    adata.obs[\"braak\"].fillna(\"Reference\", inplace=True)\n",
    "    \n",
    "    adata.obs[\"thal\"].cat.add_categories(\"Reference\", inplace=True)\n",
    "    adata.obs[\"thal\"].fillna(\"Reference\", inplace=True)\n",
    "\n",
    "    adata.obs[\"cscore\"].cat.add_categories(\"Reference\", inplace=True)\n",
    "    adata.obs[\"cscore\"].fillna(\"Reference\", inplace=True)\n",
    "    \n",
    "    adata.obs[\"lewybodydisease\"].cat.add_categories(\"Reference\", inplace=True)\n",
    "    adata.obs[\"lewybodydisease\"].fillna(\"Reference\", inplace=True)\n",
    "    \n",
    "    adata.obs[\"late_stage\"].cat.add_categories(\"Reference\", inplace=True)\n",
    "    adata.obs[\"late_stage\"].fillna(\"Reference\", inplace=True)\n",
    "\n",
    "    adata.obs[\"micro_microvascularbrain_binned\"] = \"Reference\"\n",
    "    adata.obs.loc[adata.obs[\"reference_cell\"] == 0, \"micro_microvascularbrain_binned\"] = [str(i) for i in pd.cut(adata[adata.obs[\"reference_cell\"] == 0].obs[\"micro_microvascularbrain\"], bins=3)]\n",
    "    adata.obs[\"micro_microvascularbrain_binned\"] = adata.obs[\"micro_microvascularbrain_binned\"].astype(\"category\")\n",
    "    adata.obs[\"micro_microvascularbrain_binned\"].cat.rename_categories(\n",
    "        {\n",
    "            \"(-0.01, 3.333]\": \"0 to 3 microinfarcts\",\n",
    "            \"(3.333, 6.667]\": \"4 to 6 microinfarcts\",\n",
    "            \"(6.667, 10.0]\": \"7 to 10 microinfarcts\"\n",
    "        },\n",
    "        inplace=True\n",
    "    )\n",
    "    \n",
    "    adata.obs[\"pmi_date_binned\"] = \"Reference\"\n",
    "    adata.obs.loc[adata.obs[\"reference_cell\"] == 0, \"pmi_date_binned\"] = [str(i) for i in pd.cut(adata[adata.obs[\"reference_cell\"] == 0].obs[\"pmi_date\"], bins=3)]\n",
    "    adata.obs[\"pmi_date_binned\"] = adata.obs[\"pmi_date_binned\"].astype(\"category\")\n",
    "    adata.obs[\"pmi_date_binned\"].cat.rename_categories(\n",
    "        {\n",
    "            \"(3.192, 5.939]\": \"3.2 to 5.9 hours\",\n",
    "            \"(5.939, 8.678]\": \"5.9 to 8.7 hours\",\n",
    "            \"(8.678, 11.417]\": \"8.7 to 11.4 hours\"\n",
    "        },\n",
    "        inplace=True\n",
    "    )\n",
    "\n",
    "    adata.obs[\"ch_apoe_four\"] = adata.obs[\"ch_apoestatuses\"].str.contains(\"4\")\n",
    "    adata.obs[\"ch_apoe_four\"].fillna(\"Reference\", inplace=True)\n",
    "    adata.obs[\"ch_apoe_four\"] = adata.obs[\"ch_apoe_four\"].astype(\"category\")\n",
    "    adata.obs[\"ch_apoe_four\"].cat.rename_categories({True: \"Y\", False: \"N\"}, inplace=True)\n",
    "    \n",
    "    adata.obs[\"ch_education_years_binned\"] = \"Reference\"\n",
    "    adata.obs.loc[adata.obs[\"reference_cell\"] == 0, \"ch_education_years_binned\"] = [str(i) for i in pd.cut(adata[adata.obs[\"reference_cell\"] == 0].obs[\"ch_education_years\"], bins=3)]\n",
    "    adata.obs[\"ch_education_years_binned\"] = adata.obs[\"ch_education_years_binned\"].astype(\"category\")\n",
    "    adata.obs[\"ch_education_years_binned\"].cat.rename_categories(\n",
    "        {\n",
    "            \"(11.991, 15.0]\": \"12 to 15 years\",\n",
    "            \"(15.0, 18.0]\": \"16 to 18 years\",\n",
    "            \"(18.0, 21.0]\": \"19 to 21 years\"\n",
    "        },\n",
    "        inplace=True\n",
    "    )\n",
    "    \n",
    "    adata.obs[\"reference_cell\"] = adata.obs[\"reference_cell\"].astype(\"category\")\n",
    "    adata.obs[\"reference_cell\"].cat.rename_categories(\n",
    "        {\n",
    "            0: \"False\",\n",
    "            1: \"True\",\n",
    "        },\n",
    "        inplace=True\n",
    "    )\n",
    "\n",
    "    dend_order = pd.read_csv(os.path.join(pwd, \"input\", region, \"cluster_order_and_colors.csv\"))\n",
    "    #adata.uns[\"Subclass_colors\"] = dend_order[\"subclass_color\"].drop_duplicates().to_numpy()\n",
    "    #adata.uns[\"Supertype_colors\"] = dend_order[\"cluster_color\"].to_numpy()\n",
    "\n",
    "    #adata.obs[\"class_scANVI\"].cat.rename_categories(\n",
    "    #    {\n",
    "    #        \"exc\": \"Neuronal: Glutamatergic\",\n",
    "    #        \"inh\": \"Neuronal: GABAergic\",\n",
    "    #        \"glia\": \"Non-neuronal and Non-neural\"\n",
    "    #    },\n",
    "    #    inplace=True\n",
    "    #)\n",
    "    class_order = [j for j in dend_order[\"class_label\"].drop_duplicates() if j in adata.obs[\"class_scANVI\"].cat.categories]\n",
    "    adata.obs[\"class_scANVI\"].cat.reorder_categories(class_order, inplace=True)\n",
    "\n",
    "    #adata.obs[\"subclass_scANVI\"].cat.rename_categories(\n",
    "    #    {\n",
    "    #        \"Lamp5_Lhx6\": \"Lamp5 Lhx6\",\n",
    "    #        \"Astro\": \"Astrocyte\",\n",
    "    #        \"Oligo\": \"Oligodendrocyte\",\n",
    "    #        \"Endo\": \"Endothelial\",\n",
    "    #        \"Micro-PVM\": \"Microglia-PVM\"\n",
    "    #    },\n",
    "    #    inplace=True\n",
    "    #)\n",
    "    class_order = [j for j in dend_order[\"subclass_label\"].drop_duplicates() if j in adata.obs[\"subclass_scANVI\"].cat.categories]\n",
    "    adata.obs[\"subclass_scANVI\"].cat.reorder_categories(class_order, inplace=True)\n",
    "    \n",
    "    class_order = [j for j in dend_order[\"cluster_label\"].drop_duplicates() if j in adata.obs[\"supertype_scANVI_leiden\"].cat.categories]\n",
    "    adata.obs[\"supertype_scANVI_leiden\"].cat.reorder_categories(class_order, inplace=True)\n",
    "\n",
    "    adata = adata[np.random.choice(adata.n_obs, adata.n_obs, replace=False)].copy()\n",
    "    \n",
    "    adata.obs[\"age_at_death_binned\"].cat.reorder_categories([\"Less than 65 years old\", \"65 to 77 years old\", \"78 to 89 years old\", \"90+ years old\"], inplace=True)\n",
    "    adata.obs[\"ch_cognitivestatus_binary\"].cat.reorder_categories([\"Reference\", \"No dementia\", \"Dementia\"], inplace=True)\n",
    "    adata.obs[\"adneurochange\"].cat.reorder_categories([\"Reference\", \"Not AD\", \"Low\", \"Intermediate\", \"High\"], inplace=True)\n",
    "    adata.obs[\"braak\"].cat.reorder_categories([\"Reference\", \"Braak 0\", \"Braak II\", \"Braak III\", \"Braak IV\", \"Braak V\", \"Braak VI\"], inplace=True)\n",
    "    adata.obs[\"thal\"].cat.reorder_categories([\"Reference\", \"Thal 0\", \"Thal 1\", \"Thal 2\", \"Thal 3\", \"Thal 4\", \"Thal 5\"], inplace=True)\n",
    "    adata.obs[\"cscore\"].cat.reorder_categories([\"Reference\", 'Absent', 'Sparse', 'Moderate', 'Frequent'], inplace=True)\n",
    "    adata.obs[\"lewybodydisease\"].cat.reorder_categories([\"Reference\", 'Not Identified (olfactory bulb not assessed)', 'Not Identified (olfactory bulb assessed)', 'Olfactory bulb only', 'Amygdala-predominant', 'Brainstem-predominant', 'Limbic (Transitional)', 'Neocortical (Diffuse)'], inplace=True)\n",
    "    adata.obs[\"late_stage\"].cat.reorder_categories([\"Reference\", 'Staging Precluded by FTLD with TDP43 or ALS/MND or TDP-43 pathology is unclassifiable', 'Not Identified', 'LATE Stage 1', 'LATE Stage 2', 'LATE Stage 3'], inplace=True)\n",
    "    adata.obs[\"micro_microvascularbrain_binned\"].cat.reorder_categories([\"Reference\", \"0 to 3 microinfarcts\", \"4 to 6 microinfarcts\", \"7 to 10 microinfarcts\"], inplace=True)\n",
    "    adata.obs[\"pmi_date_binned\"].cat.reorder_categories([\"Reference\", \"3.2 to 5.9 hours\", \"5.9 to 8.7 hours\", \"8.7 to 11.4 hours\"], inplace=True)\n",
    "    adata.obs[\"ch_apoe_four\"].cat.reorder_categories([\"Reference\", \"N\", \"Y\"], inplace=True)\n",
    "    adata.obs[\"ch_education_years_binned\"].cat.reorder_categories([\"Reference\", \"12 to 15 years\", \"16 to 18 years\", \"19 to 21 years\"], inplace=True)\n",
    "\n",
    "    # Keep only display metadata\n",
    "    adata.obs = adata.obs.loc[\n",
    "        :,\n",
    "        [\n",
    "            \"assay_ontology_term_id\",\n",
    "            \"suspension_type\",\n",
    "            \"cell_type_ontology_term_id\",\n",
    "            \"development_stage_ontology_term_id\",\n",
    "            \"disease_ontology_term_id\",\n",
    "            \"self_reported_ethnicity_ontology_term_id\",\n",
    "            \"organism_ontology_term_id\",\n",
    "            \"sex_ontology_term_id\",\n",
    "            \"tissue_ontology_term_id\",\n",
    "            \"is_primary_data\",\n",
    "            \"donor_id\",\n",
    "            \"reference_cell\",\n",
    "            \"class_scANVI\",\n",
    "            \"subclass_scANVI\",\n",
    "            \"supertype_scANVI_leiden\",\n",
    "            \"age_at_death_binned\",\n",
    "            \"ch_education_years_binned\",\n",
    "            \"ch_cognitivestatus_binary\",\n",
    "            \"adneurochange\",\n",
    "            \"braak\",\n",
    "            \"thal\",\n",
    "            \"cscore\",\n",
    "            \"ch_apoe_four\",\n",
    "            \"lewybodydisease\",\n",
    "            \"late_stage\",\n",
    "            \"micro_microvascularbrain_binned\",\n",
    "            \"sample_name\",\n",
    "            \"donor_name\",\n",
    "            \"pmi_date_binned\",\n",
    "            \"nCount_RNA\",\n",
    "            \"nFeature_RNA\",\n",
    "            \"fraction_mito\"\n",
    "        ]\n",
    "    ]\n",
    "    adata.obs.rename(\n",
    "        {\n",
    "            \"reference_cell\": \"Neurotypical reference\",\n",
    "            \"age_at_death_binned\": \"Age at death\",\n",
    "            \"ch_cognitivestatus_binary\": \"Cognitive status\",\n",
    "            \"adneurochange\": \"ADNC\",\n",
    "            \"braak\": \"Braak stage\",\n",
    "            \"thal\": \"Thal phase\",\n",
    "            \"cscore\": \"CERAD score\",\n",
    "            \"lewybodydisease\": \"Lewy body disease pathology\",\n",
    "            \"late_stage\": \"LATE-NC stage\",\n",
    "            \"ch_apoe_four\": \"APOE4 status\",\n",
    "            \"class_scANVI\": \"Class\",\n",
    "            \"subclass_scANVI\": \"Subclass\",\n",
    "            \"supertype_scANVI_leiden\": \"Supertype\",\n",
    "            \"sample_name\": \"Specimen ID\",\n",
    "            \"donor_name\": \"Donor ID\",\n",
    "            \"micro_microvascularbrain_binned\": \"Microinfarct pathology\",\n",
    "            \"pmi_date_binned\": \"PMI\",\n",
    "            \"ch_education_years_binned\": \"Years of education\",\n",
    "            \"nCount_RNA\": \"Number of UMIs\",\n",
    "            \"nFeature_RNA\": \"Genes detected\",\n",
    "            \"fraction_mito\": \"Fraction mitochrondrial UMIs\"\n",
    "        },\n",
    "        axis=1,\n",
    "        inplace=True\n",
    "    )\n",
    "    metadata_colors = {}\n",
    "    for i in metadata_to_color:\n",
    "        adata.obs[i] = adata.obs[i].astype(\"category\")\n",
    "        a = adata.obs[i].cat.categories.to_list()\n",
    "        cmap = plt.get_cmap('viridis', len(a)+1)\n",
    "        b = [colors.rgb2hex(cmap(j)) for j in range(cmap.N)]\n",
    "        b.pop(0)\n",
    "        if i == \"sex_ontology_term_id\":\n",
    "            b = [\"pink\", \"dodgerblue\"]\n",
    "        for A, B in zip(a, b):\n",
    "            try:\n",
    "                metadata_colors[i][A] = B\n",
    "            except:\n",
    "                metadata_colors[i] = {}\n",
    "                metadata_colors[i][A] = B\n",
    "        try:\n",
    "            del adata.uns[i + \"_colors\"]\n",
    "        except:\n",
    "            pass\n",
    "        sc.pl.umap(adata,\n",
    "            color=i,\n",
    "            title=\"\",\n",
    "            palette=metadata_colors[i],\n",
    "            size=10,\n",
    "            legend_fontsize=12\n",
    "        )\n",
    "        if i.endswith(\"_ontology_term_id\"):\n",
    "            adata.uns[i.replace(\"_ontology_term_id\", \"\") + \"_colors\"] = adata.uns[i + \"_colors\"].copy()\n",
    "            \n",
    "    tmp = dend_order.loc[:, [\"subclass_label\", \"subclass_color\"]].drop_duplicates()\n",
    "    tmp.index = tmp[\"subclass_label\"].copy()\n",
    "    tmp.drop(\"subclass_label\", axis=1, inplace=True)\n",
    "    tmp = tmp.to_dict()[\"subclass_color\"]\n",
    "    sc.pl.umap(adata,\n",
    "        color=\"Subclass\",\n",
    "        title=\"\",\n",
    "        palette=tmp,\n",
    "        size=10,\n",
    "        legend_fontsize=12\n",
    "    )\n",
    "    \n",
    "    tmp = dend_order.loc[:, [\"cluster_label\", \"cluster_color\"]].drop_duplicates()\n",
    "    tmp.index = tmp[\"cluster_label\"].copy()\n",
    "    tmp.drop(\"cluster_label\", axis=1, inplace=True)\n",
    "    tmp = tmp.to_dict()[\"cluster_color\"]\n",
    "    sc.pl.umap(adata,\n",
    "        color=\"Supertype\",\n",
    "        title=\"\",\n",
    "        palette=tmp,\n",
    "        size=10,\n",
    "        legend_fontsize=12\n",
    "    )\n",
    "    \n",
    "    for i in adata.uns.keys():\n",
    "        if i.endswith(\"_colors\"):\n",
    "            adata.uns[i] = [re.sub(\"ff$\", \"\", j) for j in adata.uns[i]]\n",
    "    \n",
    "    adata.write(os.path.join(pwd, \"output\", region, split_value.replace(\"/\", \" \") + \"_CZI_display.\" + date_tag + \".h5ad\"), compression=\"gzip\")\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "570e72c5-6bc9-420d-bd71-0037ee48f013",
   "metadata": {},
   "source": [
    "### Prepare and export objects for CZI"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ab68c553-347d-4dca-b216-02b5db43a1d2",
   "metadata": {},
   "outputs": [],
   "source": [
    "region = \"DLPFC\"\n",
    "date = \"2023-07-12\"\n",
    "adata = sc.read_h5ad(os.path.join(pwd, \"input\", region, \"final.\" + date + \".h5ad\"))\n",
    "\n",
    "adata = adata[adata.obs[\"roi\"] == \"Human DFC_all_layers\"].copy()\n",
    "adata.obs[\"supertype_scANVI\"] = adata.obs[\"supertype_scANVI\"].astype(\"object\")\n",
    "adata.obs.loc[(adata.obs[\"subclass_scANVI\"] == \"Endothelial\"), \"supertype_scANVI\"] = \"Endo_2\"\n",
    "adata.obs[\"supertype_scANVI\"] = adata.obs[\"supertype_scANVI\"].astype(\"category\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b4ef716f-6dc4-4281-a3d5-57e079d03856",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "metadata_to_color = [\n",
    "    \"sex_ontology_term_id\",\n",
    "    \"Age at death\",\n",
    "    \"Cognitive status\",\n",
    "    \"ADNC\",\n",
    "    \"Braak stage\",\n",
    "    \"Thal phase\",\n",
    "    \"CERAD score\",\n",
    "    \"Lewy body disease pathology\",\n",
    "    \"LATE-NC stage\",\n",
    "    \"APOE4 status\",\n",
    "    \"Microinfarct pathology\",\n",
    "    \"PMI\",\n",
    "    \"Years of education\"\n",
    "]\n",
    "split_key = \"subclass_scANVI\"\n",
    "\n",
    "Parallel(n_jobs=1)(\n",
    "    delayed(prepare_anndata)(\n",
    "        adata=adata[(adata.obs[split_key] == i)],\n",
    "        region=region,\n",
    "        split_key=split_key,\n",
    "        split_value=i,\n",
    "        date_tag=date,\n",
    "        metadata_to_color=metadata_to_color) for i in adata.obs[split_key].cat.categories\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "54295d2f-70d1-4f26-a0b8-7e7463d321b4",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [],
   "source": [
    "metadata_to_color = [\n",
    "    \"sex_ontology_term_id\",\n",
    "    \"Age at death\",\n",
    "    \"Cognitive status\",\n",
    "    \"ADNC\",\n",
    "    \"Braak stage\",\n",
    "    \"Thal phase\",\n",
    "    \"CERAD score\",\n",
    "    \"Lewy body disease pathology\",\n",
    "    \"LATE-NC stage\",\n",
    "    \"APOE4 status\",\n",
    "    \"Microinfarct pathology\",\n",
    "    \"PMI\",\n",
    "    \"Years of education\"\n",
    "]\n",
    "\n",
    "prepare_anndata(\n",
    "    adata=adata,\n",
    "    region=region,\n",
    "    split_key=None,\n",
    "    split_value=\"Whole Taxonomy\",\n",
    "    date_tag=\"2023-07-12\",\n",
    "    metadata_to_color=metadata_to_color\n",
    ")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
